File: server/src/Envelope.kt
package ghostbe.server

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

private val json = Json {
    encodeDefaults = true
    ignoreUnknownKeys = true
}

@Serializable
data class RequestEnvelope(
    val method: String,
    val url: String,
    val headers: Map<String, String>,
    val body: String? = null
) {
    companion object {
        fun fromJson(text: String): RequestEnvelope = json.decodeFromString(text)
    }
}

@Serializable
sealed interface ResponseEnvelope {
    @Serializable
    data class Mock(
        val intercept: Boolean = true,
        val status: Int,
        val headers: Map<String, String>,
        val body: String
    ) : ResponseEnvelope

    @Serializable
    data class Passthrough(val intercept: Boolean = false) : ResponseEnvelope
}

fun ResponseEnvelope.toJson(): String = when (this) {
    is ResponseEnvelope.Mock -> json.encodeToString(this)
    is ResponseEnvelope.Passthrough -> json.encodeToString(this)
}

File: client-ios/GhostBe/Sources/GhostBe/Envelope.swift
import Foundation

struct RequestEnvelope: Encodable {
    let method: String
    let url: String
    let headers: [String: String]
    let body: String?
}

enum ResponseEnvelope: Decodable {
    case mock(status: Int, headers: [String: String], body: String)
    case passthrough

    private enum CodingKeys: String, CodingKey { case intercept, status, headers, body }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let intercept = try container.decodeIfPresent(Bool.self, forKey: .intercept) ?? false
        if intercept {
            self = .mock(
                status: try container.decode(Int.self, forKey: .status),
                headers: try container.decode([String: String].self, forKey: .headers),
                body: try container.decode(String.self, forKey: .body)
            )
        } else {
            self = .passthrough
        }
    }
}
